src/BaGet.GCP/Services/GoogleBucketStorageService.cs

```diff
@@ -0,0 +1,85 @@
+using System;
+using System.IO;
+using System.Net;
+using System.Threading;
+using System.Threading.Tasks;
+using BaGet.Core.Services;
+using BaGet.GCP.Configuration;
+using Google;
+using Google.Cloud.Storage.V1;
+using Microsoft.Extensions.Options;
+
+namespace BaGet.GCP.Services
+{
+    public class GoogleBucketStorageService : IStorageService
+    {
+        private readonly string _bucketName;
+
+        public GoogleBucketStorageService(IOptionsSnapshot<GoogleBucketStorageOptions> options)
+        {
+            if (options == null)
+                throw new ArgumentNullException(nameof(options));
+
+            _bucketName = options.Value.BucketName;
+        }
+
+        public async Task<Stream> GetAsync(string path, CancellationToken cancellationToken = default)
+        {
+            using (var storage = await StorageClient.CreateAsync())
+            {
+                var stream = new MemoryStream();
+                await storage.DownloadObjectAsync(_bucketName, CoercePath(path), stream, cancellationToken: cancellationToken);
+                stream.Seek(0, SeekOrigin.Begin);
+                return stream;
+            }
+        }
+
+        public Task<Uri> GetDownloadUriAsync(string path, CancellationToken cancellationToken = default)
+        {
+            return Task.FromResult(new Uri($"http://storage.googleapis.com/{_bucketName}/{CoercePath(path).TrimStart('/')}"));
+        }
+
+        public async Task<PutResult> PutAsync(string path, Stream content, string contentType, CancellationToken cancellationToken = default)
+        {
+            using (var storage = await StorageClient.CreateAsync())
+            {
+                try
+                {
+                    await storage.GetObjectAsync(_bucketName, CoercePath(path), cancellationToken: cancellationToken);
```

Comment:
This isn't quite right. The steps should be something like:

1. Attempt to `UploadObjectAsync`
2. If the upload fails because the file exists (likely a `HttpStatusCode.Conflict`, but please verify this as I've never used GCP), then:
    1. Download the object
    2. Return `PutResult.AlreadyExists` if the object is the same as `content`, or `PutResult.Conflict` otherwise.

For an example of this, please see Azure's [`BlobStorageService`](https://github.com/loic-sharma/BaGet/blob/master/src/BaGet.Azure/BlobStorageService.cs#L68)